multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/util/ApplicationDigestCalculator.java

```diff
@@ -0,0 +1,106 @@
+                                                                                      .name(archiveEntryWithStreamPositions.getName())
+                                                                                      .spaceGuid(applicationArchiveContext.getSpaceId())
+                                                                                      .maxFileSizeInBytes(applicationArchiveContext.getMaxSizeInBytes())
+                                                                                      .build(),
+                                                          archiveEntryWithStreamPositions, (bytesBuffer, bytesRead) -> {
+                                                              applicationArchiveContext.calculateCurrentSizeInBytes(bytesRead);
+                                                              applicationDigestCalculator.updateDigest(bytesBuffer, 0, bytesRead);
+                                                          });
+        }
+    }
+
+    private void calculateDigestFromDirectory(ApplicationArchiveContext applicationArchiveContext, InputStream archiveStream)
+        throws IOException {
+        try (ZipArchiveInputStream zipArchiveInputStream = new ZipArchiveInputStream(archiveStream)) {
+            String moduleFileName = applicationArchiveContext.getModuleFileName();
+            ZipEntry zipEntry = applicationArchiveIterator.getFirstZipEntry(applicationArchiveContext.getModuleFileName(),
+                                                                            zipArchiveInputStream);
+            do {
+                if (!zipEntry.isDirectory()) {
+                    calculateDigestFromArchive(applicationArchiveContext, zipArchiveInputStream);
+                }
+            } while ((zipEntry = applicationArchiveIterator.getNextEntryByName(moduleFileName, zipArchiveInputStream)) != null);
+        }
+    }
+
+    private void calculateDigestFromArchive(ApplicationArchiveContext applicationArchiveContext, InputStream inputStream)
+        throws IOException {
+        byte[] buffer = new byte[BUFFER_SIZE];
+        int numberOfReadBytes = 0;
+        long maxSizeInBytes = applicationArchiveContext.getMaxSizeInBytes();
+        DigestCalculator applicationDigestCalculator = applicationArchiveContext.getDigestCalculator();
+        while ((numberOfReadBytes = inputStream.read(buffer)) != -1) {
+            long currentSizeInBytes = applicationArchiveContext.getCurrentSizeInBytes();
+            if (currentSizeInBytes + numberOfReadBytes > maxSizeInBytes) {
+                throw new ContentException(org.cloudfoundry.multiapps.mta.Messages.ERROR_SIZE_OF_FILE_EXCEEDS_CONFIGURED_MAX_SIZE_LIMIT,
+                                           currentSizeInBytes + numberOfReadBytes,
+                                           applicationArchiveContext.getModuleFileName(),
+                                           maxSizeInBytes);
```

Comment:
if we throw an exception will the inputstream be closed

Reply:
yes, it will be closed in try-with-resources